mining.go

```diff
@@ -695,7 +709,7 @@ mempoolLoop:
 				// one.
 				item := e.Value.(*txPrioItem)
 				delete(item.dependsOn, *tx.Sha())
-				if len(item.dependsOn) == 0 {
+				if len(item.dependsOn) == 0 && !priorityQueue.Contains(item) {
```

Comment:
After thinking about this a bit more, I think it would be even better to simply stop duplicate dependencies from ever being added to the dependency list anyways.  That would avoid the need for allocating a bunch of new maps here in this code.

Since the list of dependencies for a given transaction are extremely unlikely to get massively large, unlike the entire priority queue, I believe an iterative approach like you have in this PR will work well and avoid the need for extra map allocations.

The following diff would approach it that way:

```diff
$ git diff
diff --git a/mining.go b/mining.go
index ccf3d34..55d11bb 100644
--- a/mining.go
+++ b/mining.go
@@ -181,6 +181,18 @@ type BlockTemplate struct {
        ValidPayAddress bool
 }

+// containsPrioItem returns whether or not the passed list of dependcies
+// contains the transaction associated with the passed priority queue item.
+func containsPrioItem(deps *list.List, item *txPrioItem) bool {
+       for e := deps.Front(); e != nil; e = e.Next() {
+               listItem := e.Value.(*txPrioItem)
+               if *listItem.tx.Sha() == *item.tx.Sha() {
+                       return true
+               }
+       }
+       return false
+}
+
 // mergeUtxoView adds all of the entries in view to viewA.  The result is that
 // viewA will contain all of its original entries plus all of the entries
 // in viewB.  It will replace any entries in viewB which also exist in viewA
@@ -505,7 +517,9 @@ mempoolLoop:
                                        depList = list.New()
                                        dependers[*originHash] = depList
                                }
-                               depList.PushBack(prioItem)
+                               if !containsPrioItem(depList, prioItem) {
+                                       depList.PushBack(prioItem)
+                               }
                                if prioItem.dependsOn == nil {
                                        prioItem.dependsOn = make(
                                                map[wire.ShaHash]struct{})
```